rasa/core/policies/rule_policy.py

```diff
@@ -74,6 +80,154 @@ def _create_feature_key(self, states: List[Dict]) -> Text:
+        )
+
+    @staticmethod
+    def _modified_states(
+        states: List[Dict[Text, float]]
+    ) -> List[Optional[Dict[Text, float]]]:
+        """Modify the states to
+            - capture previous meaningful action before action_listen
+            - ignore previous intent
+        """
+        if states[0] is None:
+            action_before_listen = None
+        else:
+            action_before_listen = {
+                state_name: prob
+                for state_name, prob in states[0].items()
+                if PREV_PREFIX in state_name and prob > 0
+            }
+        # add `prev_...` to show that it should not be a first turn
+        indicator = PREV_PREFIX + RULE_SNIPPET_ACTION_NAME
+        if indicator not in action_before_listen.keys():
+            return [{indicator: 1}, action_before_listen, states[-1]]
+        return [action_before_listen, states[-1]]
+
+    @staticmethod
+    def _clean_feature_keys(lookup: Optional[Dict]) -> Optional[Dict]:
+        # remove action_listens that were added after conditions
+        updated_lookup = lookup.copy()
+        for key in lookup.keys():
+            # Delete rules if there is no prior action or if it would predict
+            # the `...` action
+            if PREV_PREFIX not in key or lookup[key] == RULE_SNIPPET_ACTION_NAME:
+                del updated_lookup[key]
+            elif RULE_SNIPPET_ACTION_NAME in key:
+                # If the previous action is `...` -> remove any specific state
+                # requirements for that state (anything can match this state)
+                new_key = re.sub(rf".*{PREV_PREFIX}\.\.\.[^|]*", "", key)
+
+                if new_key:
+                    if new_key.startswith("|"):
+                        new_key = new_key[1:]
+                    if new_key.endswith("|"):
+                        new_key = new_key[:-1]
+                    updated_lookup[new_key] = lookup[key]
+
+                del updated_lookup[key]
+
+        return updated_lookup
+
+    def _create_negative_lookup_from_states(
+        self,
+        trackers_as_states: List[List[Dict]],
+        trackers_as_actions: List[List[Text]],
+    ) -> Dict[Text, Text]:
+        negative_lookup = {}
+        for states, actions in zip(trackers_as_states, trackers_as_actions):
+            active_form = self._get_active_form_name(states[-1])
+            # even if there are two identical feature keys
+            # their form will be the same
+            # because of `active_form_...` feature
+            if active_form:
+                # leave only last 2 dialogue turns
+                states = self._modified_states(states[-2:])
+                feature_key = self._create_feature_key(states)
+
+                if (
+                    # form is predicted after action_listen in unhappy path,
+                    # therefore no validation is needed
+                    self._prev_action_listen_in_state(states[-1])
```

Comment:
That's only the case when the form rejected because `tracker.applied_events()` filters out the happy paths, correcT?

Reply:
this is during training so all the actions inside form are unhappy paths